% Homework template from https://www.overleaf.com/4061139pdprdn#/11876250/
\documentclass[12pt]{article}
 \usepackage[margin=1in]{geometry} 
\usepackage{amsmath,amsthm,amssymb,amsfonts}
\usepackage{enumerate}
\usepackage{url}

\usepackage{graphicx} % for \includegraphics
\usepackage{hyperref} % for hyperref urls
\hypersetup{
    colorlinks=true,
    linkcolor=blue,
    filecolor=magenta,      
    urlcolor=black,
    citecolor=blue
}

%\usepackage{natbib} % just grabbed some random bibliography package
 
\newcommand{\N}{\mathbb{N}}
\newcommand{\Z}{\mathbb{Z}}

\usepackage{float}  %For [H]
\floatplacement{figure}{H}  % default figure placement
\floatplacement{table}{H}  % default table placement
\usepackage{caption}
\usepackage{subcaption}
 
\newenvironment{problem}[2][Problem]{\begin{trivlist}
\item[\hskip \labelsep {\bfseries #1}\hskip \labelsep {\bfseries #2.}]}{\end{trivlist}}
%If you want to title your bold things something different just make another thing exactly like this but replace "problem" with the name of the thing you want, like theorem or lemma or whatever
 
 
\begin{document}

 
\title{Mining sub-graphs from gene expression partial correlations}
\author{Janet Matsen}
\maketitle

\section{Introduction} %The description of the problem you are trying to solve

My lab studies interactions between methane oxidizing microbes and other bacteria, using microbe-rich sediment at the bottom of Lake Washington as a model ecosystem. 
Understanding how these communities oxidize methane would provide insight into a major greenhouse gas mitigation system. 
Which organisms tend to pair together in nature, and what metabolic roles do each play?  
When a scoop of Lake Washington sediment is incubated with methane as the only carbon source, the population becomes enriched in methane-consuming methanotrophs.
What is more surprising is the consistent appearance of non-methanotrophic "methylotrophic" bacteria.
These microbes can use single carbon-compounds other than methane, such as methanol and methylamines. 
A third group also appears; these microbes presumably only eat multi-carbon compounds.

It is suspected that the methanotrophs are providing the methylotrophs with methanol, but are there additional nutrient exchanges?
Perhaps the methylotrophs are assisting with nitrogen fixation, or synthesis of key nutrients. 
What exactly is the third group is consuming?  
Are they eating other secretions or just scavenging the byproducts of cell death?
Are they actively killing the methanotrophs and methylotrophs?

Our lab has gene expression profiles from 83 samples of sediment incubation. 
From these, a partial correlation matrix was previously computed. 
Such a partial correlation matrix can be represented as a graph and mined to identify sub-graph motifs demonstrating metabolic cooperation between microbes. 
In particular, identification of hubs and cycles that include genes from two bacteria can be used to discover metabolic cooperation, such as nutrient exchange. 
All code for this project is at \href{https://github.com/JanetMatsen/Neo4j_meta4}{github.com/JanetMatsen/Neo4j\_meta4}.

%This project began by 
%This project aimed to infer specific metabolic interactions between microbes in these samples using graph algorithms and queries.
% In addition, it provides a pilot study for the suitibility of Neo4j for future network-based research in our lab.

% Hubs are small sub-graphs that are more highly connected than the background. 

%We are motivated to find patterns that illustrate metabolic cooperation between different organisms in our samples.  
%For example, we would expect to find sub-graphs demonstrating metabolic cooperation between methane-consuming methanotrophs and methanol-consuming methylotrophs.  
%There are many methanotrophs and many methylotrophs in our samples, and thus many possible pairs to cooperate metabolically.  
%Searching our graph for similar motifs that include both methanotrophs and methylotrophs could uncover which pairs tend to interact, and what genes' expression are linked to these co-occurrences.  
%These motifs are directly useful for our lab, as they lead to testable hypotheses for wet lab experiments.

\section{Related Work}
%A description of related work. How does your approach compare with what has been done before. Suggestion: cite about 3 papers. Don't worry about finding *all* the existing related work. Our worry here is not the novelty of your project. The goal is for you to learn that placing your work in perspective of what has been done before is a crucial step in research. For the purpose of the class, we expect your related work to be incomplete.

Graphs are a useful tool for describing how species interact \cite{borthagaray2014inferring}.
Partial correlations are often used to determine the connections in a graph in which the edges represent influence of one node on another \cite{hero2012hub}. 
Once a graph is in hand, we can find community structure: groups of nodes where the nodes within a group are highly connected to each other but there are fewer connections between groups \cite{girvan2002community}.
Recent work has applied partial correlation to the ``phyllosphere'' microbiome, the collection of bacteria that live on plants \cite{agler2016microbial}.

%\cite{Mateos:2011aa}

\section{Methods} %The description of your approach

\subsection{Choice of Neo4j}

Neo4j was chosen for its user-friendly query language, Cypher. % TODO: cite
While the queries required to execute this project are no sophisticated, downstream data analysis and exploration will be greatly accelerated by the ability to perform queries on the graph and display them in Neo4j's browser. 
For example, we may wish to query for a chain of nodes such as \texttt{(metabolite production organism:A) --> (metabolite export organism:A) --> (metabolite import organism:B) --> (metabolite use organism:B)}.  

Neo4j does have drawbacks that did not affect this project but should be considered for larger scale projects.
First, though multiple clients can run in parallel on the same database, Neo4j cannot scale queries across multiple servers.  
When the query run time is a key concern, tools such as Grail and Giraph \cite{fan2015case} should be considered. 
Grail provides syntactic layers that allow user-friendly graph-query like APIs function on traditional RDBMSs.
% The Case Against Specialized Graph Analytics Engines. CIDR 2015
Giraph focuses on parallelizability by using Hadoop. 
Myria also enhances parallelizability and graph-like queries with the advantage of handling distribution of the database across remote machines. 

\subsection{Graph construction and connected components}

The partial correlation network was modeled Neo4j, the leading graph database.
Nodes represent genes, and edges connect genes with statistically significant partial correlation with one another. 
The nodes are labeled with the species, gene id, and protein name.
Edges are labeled with the significance weight, and sign (positive or negative). 
Two gene's nodes have an edge with a large positive value when those two genes tend to be highly expressed together throughout our samples. 
Conversely, large negative edge weights indicate that the corresponding genes tend to be anticorrelated. 

I prepared an information-enriched dataset for import into Neo4j from a file listing only gene loci and a partial correlation value (see \href{https://github.com/JanetMatsen/Neo4j_meta4/blob/master/jupyter/prepare_whole_network.ipynb}{this notebook}).  
The result was a 50 million row tsv file enriched with genome names, gene products, magnitudes of partial correlations, and a label for cross-species partial correlations.

Different databases were built for varying partial correlation values in the set \texttt{[0.005,  0.0075,  0.01,  0.0125,  0.015,  0.02,  0.025,  0.03,  0.035,  0.04,  0.045,  0.05,  0.055,  0.06]}
The following \href{https://github.com/JanetMatsen/Neo4j_meta4/blob/master/data_mining_Neo4j_v2_3_2/queries/load_network--specify_cutoff.txt}{Cypher query} was used to iterate over the 50 million edge dataset and build databases:

\begin{verbatim}
        LOAD CSV WITH HEADERS FROM
            'file: \%s'
        AS line FIELDTERMINATOR '\t'
        WITH toFloat(line.pcor) AS pcor,
             abs(toFloat(line.pcor)) AS abs_pcor,
             line.cross_species AS cross_species,
             line.association AS association,
             line.source_locus_tag AS source_locus_tag,
             line.target_locus_tag AS target_locus_tag,
             line.source_organism_name AS source_organism_name,
             line.target_organism_name AS target_organism_name,
             line.source_gene AS source_gene,
             line.target_gene AS target_gene,
             line.source_gene_product AS source_gene_product,
             line.target_gene_product AS target_gene_product
        WHERE abs_pcor > %f
        MERGE (g1:Gene {locus_tag:source_locus_tag,
                        organism:source_organism_name,
                        gene:source_gene,
                        gene_product:source_gene_product})
        MERGE (g2:Gene {locus_tag:target_locus_tag,
                        organism:target_organism_name,
                        gene:target_gene,
                        gene_product:target_gene_product})
        MERGE (g1) -[:X {pcor:pcor,
                         pcor_abs:abs_pcor,
                         cross_species:cross_species,
                         association:association}]-> (g2)
\end{verbatim}

The \texttt{MERGE} operator ensures the pattern exists; if it does not find a piece of the pattern, it adds the corresponding nodes and edges. 
The \texttt{\%s} at the top and \texttt{\%f} in the \texttt{WHERE} clause allowed string substitution in Java to specify which source file and cutoff to use. 
I used the Neo4j Java API to call this query from Eclipse (see \href{https://github.com/JanetMatsen/Neo4j_meta4/blob/master/data_mining_Neo4j_v2_3_2/src/ConstructNetwork50M.java}{ConstructNetwork50M.java}), and count the number of nodes added to the database. 
After my script worked well, I added command line arguments and saved the script to a Jar to be run on a remote machine with large memory.

I made a second Jar (see \href{https://github.com/JanetMatsen/Neo4j_meta4/blob/master/data_mining_Neo4j_v2_3_2/src/ConnectedComponentsFinder50M.java}{ConnectedComponentsFinder50M.java}) that runs \href{https://github.com/besil/Neo4jSNA}{Neo4jSNA}'s \footnote{\url{https://github.com/besil/Neo4jSNA}} connected components algorithm.
Neo4jSNA was most recently updated for Neo4j 2.3.2, which is much older than the current version of Neo4j, which is 3.0.7. 
Even the latest 2-series version of Neo4j, 2.3.7, is not compatible with Neo4jSNA, but the JARs for 2.3.2 was found via a link from the Chocolatey package manager.
The connected components algorithm labeled the nodes in my Neo4j database with the corresponding connected components identification number.
The labels enumerated by iterating over notes using the Java API.

These jars were each called by Python (see \href{https://github.com/JanetMatsen/Neo4j_meta4/blob/master/code/database_comparisons.py}{database\_comparisons.py} and the corresponding \href{https://github.com/JanetMatsen/Neo4j_meta4/blob/master/jupyter/build_graphs_and_find_connected_components_50M.ipynb}{ipython notebook}).
I had python objects to represent information at each level of the analysis.  
The \href{https://github.com/JanetMatsen/Neo4j_meta4/blob/master/code/database_comparisons.py}{Database} class could submit jobs to create single databases and parse statistics about the product from Java's standard output.
The \href{https://github.com/JanetMatsen/Neo4j_meta4/blob/master/code/database_comparisons.py}{DatabaseComparison} class contained many database objects, and included many cross-database plotting methods.
Database objects each contained a \texttt{connected\_components} attribute, which was an instance of my \href{https://github.com/JanetMatsen/Neo4j_meta4/blob/master/code/connected_component.py}{ConnectedComponentsDB} class.
A single database's ConnectedComponentsDB included one or more \href{https://github.com/JanetMatsen/Neo4j_meta4/blob/master/code/connected_component.py}{ConnectedComponent} instances.
These Python structures allowed me to explore results across different granularities in my jupyter notebook sessions.

\section{Results}  

Queries ran fast. 

\begin{figure}[H]
    \captionsetup{width=0.6\textwidth}
    \centering
    \includegraphics[scale=0.8]{../../figures_curated/161209_hist_of_pcor_values_50_bins.pdf}
    \caption{Distribution of partial correlation values each database was built from.}
    %\label{fig:ABCD}
\end{figure}

\begin{figure}[H]
    \captionsetup{width=0.6\textwidth}
    \centering
    \includegraphics[scale=0.55]{../../figures_curated/{161213_heatmap_cutof_0.02}.png}
    \caption{Distribution of partial correlation values each database was built from.}
    %\label{fig:ABCD}
\end{figure}


These will be filtered based on the most essential feature: a mixture of organism types including one methanotroph and one methylotroph.
If these connected components are very large, I will seek to break them apart by filtering for magnitude of covariance, and look for cliques within. 
Then I will use techniques in from social network analysis to do community detection. 

I will leverage existing packages when available, and write my own algorithms when necessary. 




\section{Future work}

Connected components $\rightarrow$ hubs and cycles. 

Coloring by organism type or partial correlation sign requires making each node a different type; requires more sophisticated database construction. 


Third, if I have time, I will try to implement the community detection algorithm described in \cite{girvan2002community}.
This algorithm works to find edges that are least central to communities in the graph and removes them.
Such ``between'' edges can be found by looking for edges that are often used in shortest paths between pairs of nodes.



%\bibliographystyle{te}

\bibliographystyle{plain}
\bibliography{neo4j}
 
 
\end{document}